src/app/store/reducers/mail.reducers.ts

```diff
@@ -155,12 +164,47 @@ export function reducer(state = initialState, action: MailActions): MailState {
       };
     }
 
+    case MailActionTypes.UPLOAD_ATTACHMENT_REQUEST: {
+      state.attachments.forEach((item, index) => {
+        if (item.attachmentId === action.payload.attachmentId) {
+          state.attachments[index].request = action.payload.request;
+        }
+      });
+      return { ...state };
+    }
+
     case MailActionTypes.UPLOAD_ATTACHMENT_SUCCESS: {
+      state.attachments.forEach((item, index) => {
+        if (item.attachmentId === action.payload.data.attachmentId) {
+          if (item.hash === action.payload.response.hash) {
+            state.attachments[index].id = action.payload.response.id;
+            state.attachments[index].inProgress = false;
+            state.attachments[index].request = null;
+          }
+        }
+      });
       return {
         ...state
       };
     }
 
+    case MailActionTypes.DELETE_ATTACHMENT: {
+      const index = state.attachments.findIndex(attachment => attachment.attachmentId === action.payload.attachmentId);
+      if (index > -1 && !state.attachments[index].id) {
+        state.attachments[index].request.unsubscribe();
+        state.attachments.splice(index, 1);
+      }
+      return { ...state };
+    }
+
+    case MailActionTypes.DELETE_ATTACHMENT_SUCCESS: {
+      const index = state.attachments.findIndex(attachment => attachment.id === action.payload.id);
```

Comment:
This code can be trimmed to 
state.attachments = state.attachements.filter(attachment => attachment.id !== action.payload.id)

Enforces immutability

Reply:
No need of checking index and splicing it

Reply:
@zainzafar90 you are right.
I will ask waseem to handle it in next PR